File: architecture-skills/design-patterns/examples/builder/builder.java
// Builder pattern in Java.
//
// The classic home of the pattern: a static nested Builder producing an
// immutable product (compare java.net.http.HttpRequest.newBuilder()).

import java.util.LinkedHashMap;
import java.util.Map;

public class builder {

    /** Immutable product. */
    static final class HttpRequest {
        private final String method;
        private final String url;
        private final Map<String, String> headers;
        private final String body; // nullable
        private final double timeoutSeconds;

        private HttpRequest(Builder b) {
            this.method = b.method;
            this.url = b.url;
            this.headers = Map.copyOf(b.headers);
            this.body = b.body;
            this.timeoutSeconds = b.timeoutSeconds;
        }

        static Builder newBuilder(String method, String url) {
            return new Builder(method, url);
        }

        static final class Builder {
            private final String method;
            private final String url;
            private final Map<String, String> headers = new LinkedHashMap<>();
            private String body;
            private double timeoutSeconds = 30.0;

            private Builder(String method, String url) {
                this.method = method.toUpperCase();
                this.url = url;
            }

            Builder header(String name, String value) {
                headers.put(name, value);
                return this;
            }

            Builder body(String body) {
                this.body = body;
                return this;
            }

            Builder timeout(double seconds) {
                this.timeoutSeconds = seconds;
                return this;
            }

            HttpRequest build() {
                if (body != null && (method.equals("GET") || method.equals("HEAD"))) {
                    throw new IllegalStateException(method + " requests cannot have a body");
                }
                return new HttpRequest(this);
            }
        }

        @Override
        public String toString() {
            return method + " " + url + " (timeout " + timeoutSeconds + "s) headers=" + headers
                    + (body != null ? " body=" + body : "");
        }
    }

    public static void main(String[] args) {
        HttpRequest request = HttpRequest.newBuilder("POST", "https://api.example.com/users")
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer token123")
                .body("{\"name\": \"Ada\"}")
                .timeout(5.0)
                .build();

        System.out.println(request);
    }
}

File: architecture-skills/design-patterns/examples/builder/builder.swift
// Builder pattern in Swift.
//
// Note: memberwise initializers with default values cover simple cases,
// and result builders (@resultBuilder, as in SwiftUI) are the language-
// native form for declarative trees. Use an explicit builder for
// multi-stage construction with validation.

import Foundation

/// Immutable product.
struct HttpRequest {
    let method: String
    let url: String
    let headers: [String: String]
    let body: String?
    let timeoutSeconds: Double
}

enum BuildError: Error {
    case bodyNotAllowed(method: String)
}

final class HttpRequestBuilder {
    private let method: String
    private let url: String
    private var headers: [String: String] = [:]
    private var body: String?
    private var timeoutSeconds: Double = 30

    init(method: String, url: String) {
        self.method = method.uppercased()
        self.url = url
    }

    @discardableResult
    func header(_ name: String, _ value: String) -> HttpRequestBuilder {
        headers[name] = value
        return self
    }

    @discardableResult
    func body(_ body: String) -> HttpRequestBuilder {
        self.body = body
        return self
    }

    @discardableResult
    func timeout(_ seconds: Double) -> HttpRequestBuilder {
        timeoutSeconds = seconds
        return self
    }

    func build() throws -> HttpRequest {
        if body != nil, method == "GET" || method == "HEAD" {
            throw BuildError.bodyNotAllowed(method: method)
        }
        return HttpRequest(
            method: method,
            url: url,
            headers: headers,
            body: body,
            timeoutSeconds: timeoutSeconds
        )
    }
}

let request = try HttpRequestBuilder(method: "POST", url: "https://api.example.com/users")
    .header("Content-Type", "application/json")
    .header("Authorization", "Bearer token123")
    .body(#"{"name": "Ada"}"#)
    .timeout(5)
    .build()

print(request)
